File: src/tools/semgrep/eval-repos/synthetic/java/SecurityPatterns.java
/**
 * Test file for Java security vulnerability detection.
 * Contains patterns detected by p/java ruleset.
 */
package synthetic.java;

import java.io.*;
import java.sql.*;
import javax.servlet.http.*;
import javax.xml.parsers.*;
import org.xml.sax.InputSource;

public class SecurityPatterns {

    // SQL_INJECTION: JDBC concatenation
    public ResultSet unsafeJdbcQuery(Connection conn, String userId) throws SQLException {
        String query = "SELECT * FROM users WHERE id = " + userId;
        Statement stmt = conn.createStatement();
        return stmt.executeQuery(query);
    }

    // SQL_INJECTION: String format in query
    public ResultSet unsafeFormattedQuery(Connection conn, String name) throws SQLException {
        String query = String.format("SELECT * FROM users WHERE name = '%s'", name);
        Statement stmt = conn.createStatement();
        return stmt.executeQuery(query);
    }

    // COMMAND_INJECTION: Runtime.exec with user input
    public Process unsafeRuntimeExec(String command) throws IOException {
        Runtime runtime = Runtime.getRuntime();
        return runtime.exec(command);
    }

    // COMMAND_INJECTION: ProcessBuilder with user input
    public Process unsafeProcessBuilder(String command) throws IOException {
        ProcessBuilder pb = new ProcessBuilder(command);
        return pb.start();
    }

    // UNSAFE_DESERIALIZATION: ObjectInputStream
    public Object unsafeDeserialization(InputStream input) throws Exception {
        ObjectInputStream ois = new ObjectInputStream(input);
        return ois.readObject();
    }

    // XXE_VULNERABILITY: DocumentBuilder without security features
    public void unsafeXmlParsing(String xml) throws Exception {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        DocumentBuilder builder = factory.newDocumentBuilder();
        builder.parse(new InputSource(new StringReader(xml)));
    }

    // XSS_VULNERABILITY: Direct output to response
    public void unsafeXssOutput(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String name = req.getParameter("name");
        resp.getWriter().println("<h1>Hello " + name + "</h1>");
    }

    // INSECURE_CRYPTO: Weak random
    public int unsafeRandom() {
        java.util.Random random = new java.util.Random();
        return random.nextInt();
    }

    // INSECURE_CRYPTO: ECB mode
    public byte[] unsafeEcbMode(byte[] data, byte[] key) throws Exception {
        javax.crypto.Cipher cipher = javax.crypto.Cipher.getInstance("AES/ECB/PKCS5Padding");
        javax.crypto.spec.SecretKeySpec keySpec = new javax.crypto.spec.SecretKeySpec(key, "AES");
        cipher.init(javax.crypto.Cipher.ENCRYPT_MODE, keySpec);
        return cipher.doFinal(data);
    }

    // OPEN_REDIRECT: Redirect without validation
    public void unsafeRedirect(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String url = req.getParameter("redirect");
        resp.sendRedirect(url);
    }

    // SAFE: Parameterized query
    public ResultSet safeQuery(Connection conn, String userId) throws SQLException {
        String query = "SELECT * FROM users WHERE id = ?";
        PreparedStatement stmt = conn.prepareStatement(query);
        stmt.setString(1, userId);
        return stmt.executeQuery();
    }

    // SAFE: Secure random
    public int safeRandom() throws Exception {
        java.security.SecureRandom random = new java.security.SecureRandom();
        return random.nextInt();
    }

    // SAFE: CBC mode with IV
    public byte[] safeCbcMode(byte[] data, byte[] key, byte[] iv) throws Exception {
        javax.crypto.Cipher cipher = javax.crypto.Cipher.getInstance("AES/CBC/PKCS5Padding");
        javax.crypto.spec.SecretKeySpec keySpec = new javax.crypto.spec.SecretKeySpec(key, "AES");
        javax.crypto.spec.IvParameterSpec ivSpec = new javax.crypto.spec.IvParameterSpec(iv);
        cipher.init(javax.crypto.Cipher.ENCRYPT_MODE, keySpec, ivSpec);
        return cipher.doFinal(data);
    }
}

File: src/tools/semgrep/eval-repos/synthetic/go/security_patterns.go
// Test file for Go security vulnerability detection.
// Contains patterns detected by p/go ruleset.
package synthetic

import (
	"crypto/md5"
	"crypto/tls"
	"database/sql"
	"fmt"
	"html/template"
	"io/ioutil"
	"math/rand"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
)

// SQL_INJECTION: String concatenation in SQL query
func UnsafeSqlQuery(db *sql.DB, userId string) (*sql.Rows, error) {
	query := "SELECT * FROM users WHERE id = " + userId
	return db.Query(query)
}

// SQL_INJECTION: fmt.Sprintf in SQL query
func UnsafeFormattedQuery(db *sql.DB, name string) (*sql.Rows, error) {
	query := fmt.Sprintf("SELECT * FROM users WHERE name = '%s'", name)
	return db.Query(query)
}

// COMMAND_INJECTION: exec.Command with user input
func UnsafeExecCommand(userCommand string) ([]byte, error) {
	cmd := exec.Command("sh", "-c", userCommand)
	return cmd.Output()
}

// COMMAND_INJECTION: os.exec with shell
func UnsafeOsExec(command string) ([]byte, error) {
	return exec.Command(command).Output()
}

// XSS_VULNERABILITY: template without escaping
func UnsafeTemplateRender(w http.ResponseWriter, userInput string) {
	tmpl := template.Must(template.New("test").Parse("<h1>Hello " + userInput + "</h1>"))
	tmpl.Execute(w, nil)
}

// PATH_TRAVERSAL: filepath.Join with user input
func UnsafeFileRead(basePath, userPath string) ([]byte, error) {
	fullPath := filepath.Join(basePath, userPath)
	return ioutil.ReadFile(fullPath)
}

// PATH_TRAVERSAL: Direct file access with user input
func UnsafeDirectFileAccess(filename string) ([]byte, error) {
	return os.ReadFile(filename)
}

// INSECURE_CRYPTO: MD5 hashing
func UnsafeMd5Hash(data []byte) []byte {
	hash := md5.Sum(data)
	return hash[:]
}

// INSECURE_CRYPTO: TLS InsecureSkipVerify
func UnsafeTlsConfig() *tls.Config {
	return &tls.Config{
		InsecureSkipVerify: true,
	}
}

// INSECURE_CRYPTO: Weak random with math/rand
func UnsafeWeakRandom() int {
	return rand.Intn(100)
}

// SSRF_VULNERABILITY: HTTP request with user-controlled URL
func UnsafeSsrfRequest(userUrl string) (*http.Response, error) {
	return http.Get(userUrl)
}

// RESOURCE_LEAK: Defer in loop
func UnsafeDeferInLoop(files []string) {
	for _, f := range files {
		file, _ := os.Open(f)
		defer file.Close() // RESOURCE_LEAK: defer in loop
	}
}

// SAFE: Parameterized query
func SafeSqlQuery(db *sql.DB, userId string) (*sql.Rows, error) {
	return db.Query("SELECT * FROM users WHERE id = ?", userId)
}

// SAFE: Proper template escaping
func SafeTemplateRender(w http.ResponseWriter, userInput string) {
	tmpl := template.Must(template.New("test").Parse("<h1>Hello {{.}}</h1>"))
	tmpl.Execute(w, template.HTMLEscapeString(userInput))
}

// SAFE: Crypto-secure random
import "crypto/rand"

func SafeSecureRandom() ([]byte, error) {
	b := make([]byte, 32)
	_, err := rand.Read(b)
	return b, err
}

// SAFE: Proper TLS config
func SafeTlsConfig() *tls.Config {
	return &tls.Config{
		MinVersion: tls.VersionTLS12,
	}
}
